third_party/terraform/utils/common_polling.go

```diff
@@ -0,0 +1,82 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"reflect"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+type PollWaiter interface {
+	// Takes the response and error from a given GET request to determine
+	// whether Terraform should continue polling/waiting.
+	// continuePolling should be true if Terraform should continue polling.
+	// err should be nil if last response was successful. Otherwise, it should
+	// detail the reason for continuing polling or reason to terminate and
+	// return an error.
+	ShouldContinuePolling(resp map[string]interface{}, respErr error) (continuePolling bool, err error)
+}
+
+func PollingWaitTime(config *Config, waiter PollWaiter, project, url, activity string, timeout time.Duration) error {
+	log.Printf("[DEBUG] %s: Polling until expected state is read", activity)
+
+	return resource.Retry(timeout, func() *resource.RetryError {
+		readResp, readErr := sendRequest(config, "GET", project, url, nil)
+		pollAgain, err := waiter.ShouldContinuePolling(readResp, readErr)
+		if pollAgain {
+			log.Printf("[DEBUG] Response has not terminated, continue polling: %v", err)
+			if err == nil {
+				err = errors.New("got pending response")
+			}
+			return resource.RetryableError(err)
+		}
+		log.Printf("[DEBUG] Response has terminated with error: %v", err)
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+}
+
+// ResourceStatePoll implements PollWaiter. It should be used for resources where
+// a create/update operation returns the desired state and we want to poll
+// until GET state value converges.
+// NOTE: This comparison to desired state uses reflect.DeepEqual() and
+// thus must match exactly.
+// nolint: deadcode,unused
+type ResourceStatePollWaiter struct {
+	expected map[string]interface{}
```

Comment:
I don't really like adding this state to the waiter, especially since this choice has bled into the other waiter implementations (forcing the pollwaiter funcs to accept and ignore `res`). I also imagine that polling equality is going to be very business logic heavy per resource (ignore auto-populated timestamps, synonymous fields, etc) I don't expect that a naive resource equality is going to be usable for most resources. Given that maybe this is better expressed as a comparison method or interface passed in, similar to the [sort method](https://golang.org/pkg/sort/#Sort).

Though either way until there is something using the ResourceState poll waiter it feels like premature abstraction.

Reply:
do you think we should remove it? i switched so it just takes a func in - I personally don't care whether we leave it in or not, but i think i'd rather have it as an example of how we might in the future expand terraform yaml to take in custom poll waiter definitions. The function thing is a much better idea.